Implement employee sorting by age and name

koder95 opened the pull request:
Added a custom comparator to sort employees first by age and then by name. The getEmployByOrder method now returns a sorted set of employees using this comparator.

koder95 committed:
Implement employee sorting by age and name

Added a custom comparator to sort employees first by age and then by name. The getEmployByOrder method now returns a sorted set of employees using this comparator.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/EmployeeService.java:
```diff
@@ -1,10 +1,21 @@
+
     public Set<Employee> getEmployByOrder(List<Employee> employees) {
-        return null;
+        SortedSet<Employee> employeeSortedSet = new TreeSet<>(AGE_NAME_COMPARATOR);
```
**CRITICAL: Method name mismatch.**
The method is declared as `getEmployByOrder` (typo) instead of the required `getEmployeeByOrder(List<Employee> employees)`. Tests and the task description expect that exact name/signature. Rename the method to match the assignment contract so it will be discovered and executed by tests.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/EmployeeService.java:
```diff
@@ -1,10 +1,21 @@
+    private static final Comparator<Employee> AGE_NAME_COMPARATOR = new Comparator<Employee>() {
+        @Override
+        public int compare(Employee first, Employee second) {
+            return first == second || first.equals(second) ? 0
```
**MAJOR: Comparator relies on equals() for equality.**
The comparator returns 0 when `first.equals(second)`. The comparator must determine equality based on the specified sort keys (age and name), not on `equals()`. If `Employee.equals` is not implemented to compare age and name, distinct objects with identical age/name will not be considered duplicates by the TreeSet. Consider implementing the comparator so it returns 0 exactly when `age` and `name` are equal (e.g., compare age then name).

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/EmployeeService.java:
```diff
@@ -1,10 +1,21 @@
+            return first == second || first.equals(second) ? 0
+                    : first.getAge() > second.getAge() ? 1
+                    : first.getAge() < second.getAge() ? -1
+                    : first.getName() == null ? -1 : first.getName().compareTo(second.getName());
```
**MAJOR: Unsafe null handling for names (possible NPE).**
This line checks `first.getName() == null ? -1 : first.getName().compareTo(second.getName())` but does not check whether `second.getName()` is null. If `second.getName()` is null and `first.getName()` is non-null, `compareTo` will throw a NullPointerException. Decide how to treat null names and implement symmetric handling (for example, use `thenComparing(Comparator.nullsFirst(Comparator.naturalOrder()))`).

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/EmployeeService.java:
```diff
@@ -1,10 +1,21 @@
+import java.util.*;
 
 public class EmployeeService {
+    private static final Comparator<Employee> AGE_NAME_COMPARATOR = new Comparator<Employee>() {
```
**Suggestion: Use the fluent Comparator API for clarity and correctness.**
The nested ternary comparator is hard to read and error-prone. Prefer a clear comparator expression such as:
`Comparator<Employee> cmp = Comparator.comparingInt(Employee::getAge).thenComparing(Employee::getName);`
Add `Comparator.nullsFirst/Last` around `Employee::getName` if you need to handle null names. This will ensure the comparator defines a total order consistent with the requirements.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/EmployeeService.java:
```diff
@@ -1,10 +1,21 @@
     public Set<Employee> getEmployByOrder(List<Employee> employees) {
-        return null;
+        SortedSet<Employee> employeeSortedSet = new TreeSet<>(AGE_NAME_COMPARATOR);
+        employeeSortedSet.addAll(employees);
```
**Note: addAll without null-check on input list.**
The method calls `employeeSortedSet.addAll(employees)` without checking whether `employees` is null. If null inputs are possible in tests, this will throw. Either validate input or document that null is not allowed. If defensive coding is desired, return an empty TreeSet when `employees` is null.

mateacademy-ai-mentor reviewed (changes requested):
### Review Summary

**Decision: REQUEST_CHANGES**

Your solution is close, but there are critical blockers that will cause tests to fail and can break functionality.

Critical issues to fix:
- Method name/signature mismatch (src/main/java/core/basesyntax/EmployeeService.java): The task requires `public Set<Employee> getEmployeeByOrder(List<Employee> employees)`, but you implemented `getEmployByOrder(...)`. This prevents the tests from finding and running your method. Rename it to the exact required signature.
- Comparator equality and null-safety (same file, comparator implementation): The comparator returns `0` when `first.equals(second)` and handles null names asymmetrically. This can violate the comparator contract and cause NPEs if `second.getName()` is null. The comparator must consider two employees equal only when both `age` and `name` are equal, and it must handle null names consistently on both sides. Suggested approach: compare by age first, then by name using a null-safe comparison (e.g., `Comparator.comparingInt(Employee::getAge).thenComparing(Comparator.comparing(Employee::getName, Comparator.nullsFirst(String::compareTo)))`).

Minor improvements (optional but recommended):
- Prefer the fluent Comparator API for readability and correctness instead of nested ternaries.
- Consider a null-check for the input list before `addAll` to avoid NPE if `employees` can be null.

Please address the method name and comparator issues; after that, your approach with `TreeSet` should meet the requirements. Good use of a sorted set overall—fixes here will make it robust and test-friendly.

***

✨ Please rate this review in the Platform's notification center[^1]. I will be grateful for your feedback! ✨

[^1]: [Rate AI review example](https://snipboard.io/gD6XKF.jpg)

> [!IMPORTANT]
> Fix the issues, commit the changes, and then re-request my review using the **`Re-request review` button**[^2] in the Reviewers tab.

[^2]: [Re-request AI review example](https://snipboard.io/nOXxHN.jpg)

> [!TIP]
> If you have any comments about this review, please leave a message in the [Add a comment](#new_comment_field) section below and **mention me with the `@mateacademy-ai-mentor` tag**[^3].
> I will read your message during my next review of your pull request. Please do not reply directly to my comments, as that feature is not supported yet.

[^3]: [Add AI review comments example](https://snipboard.io/hVlOey.jpg)

> [!CAUTION]
> Do not close the pull request until the review is complete. Closing it early can lead to unexpected behavior and disrupt the review process.

koder95 committed:
Rename getEmployByOrder to getEmployeeByOrder

Renamed the method getEmployByOrder to getEmployeeByOrder in EmployeeService and updated all usages in EmployeeServiceTest for clarity and consistency.

koder95 committed:
Refactor employee comparator for clarity

Simplified the AGE_NAME_COMPARATOR using Comparator.comparingInt and thenComparing with nullsFirst for name comparison. This improves readability and maintainability of the comparator logic.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/EmployeeService.java:
```diff
@@ -1,10 +1,15 @@
 public class EmployeeService {
-    public Set<Employee> getEmployByOrder(List<Employee> employees) {
-        return null;
+    private static final Comparator<Employee> AGE_NAME_COMPARATOR = Comparator
+            .comparingInt(Employee::getAge)
+            .thenComparing(Employee::getName, Comparator.nullsFirst(String::compareTo));
```
Good: The comparator is declared using the fluent Comparator API and implements the required ordering: first by `age` ascending, then by `name` with nulls handled via `nullsFirst`. This matches the task requirement to create a custom comparator instead of relying on Comparable.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/EmployeeService.java:
```diff
@@ -1,10 +1,15 @@
+            .comparingInt(Employee::getAge)
+            .thenComparing(Employee::getName, Comparator.nullsFirst(String::compareTo));
+
+    public Set<Employee> getEmployeeByOrder(List<Employee> employees) {
```
The method signature exactly matches the required signature `Set<Employee> getEmployeeByOrder(List<Employee> employees)`. This is important because the tests call this exact method. Good.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/EmployeeService.java:
```diff
@@ -1,10 +1,15 @@
+
+    public Set<Employee> getEmployeeByOrder(List<Employee> employees) {
+        SortedSet<Employee> employeeSortedSet = new TreeSet<>(AGE_NAME_COMPARATOR);
+        employeeSortedSet.addAll(employees);
```
Potential NPE: `employeeSortedSet.addAll(employees)` will throw a NullPointerException if the caller passes `employees == null`. Consider guarding the input, for example returning an empty `TreeSet` when `employees` is null, or explicitly validating the parameter before calling `addAll`.

koder95 replied:
Will it be enought?
```suggestion
        if (employees != null) {
            employeeSortedSet.addAll(employees);
        }
```

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/EmployeeService.java:
```diff
@@ -1,10 +1,15 @@
+            .thenComparing(Employee::getName, Comparator.nullsFirst(String::compareTo));
+
+    public Set<Employee> getEmployeeByOrder(List<Employee> employees) {
+        SortedSet<Employee> employeeSortedSet = new TreeSet<>(AGE_NAME_COMPARATOR);
```
Potential runtime issue when the input list contains `null` elements: adding a `null` Employee reference into the TreeSet will cause an exception because the comparator expects non-null Employee objects. If you need to accept null elements, create the TreeSet with `Comparator.nullsFirst(AGE_NAME_COMPARATOR)` or filter out nulls before adding. If null elements are not expected, consider documenting that behavior.

mateacademy-ai-mentor commented on src/main/java/core/basesyntax/EmployeeService.java:
```diff
@@ -1,10 +1,15 @@
+    public Set<Employee> getEmployeeByOrder(List<Employee> employees) {
+        SortedSet<Employee> employeeSortedSet = new TreeSet<>(AGE_NAME_COMPARATOR);
+        employeeSortedSet.addAll(employees);
+        return employeeSortedSet;
```
Return is correct: returning the populated SortedSet ensures the iteration order reflects the comparator. No change needed here unless you implement the defensive checks above.

mateacademy-ai-mentor commented on src/test/java/core/basesyntax/EmployeeServiceTest.java:
```diff
@@ -33,7 +33,7 @@ public static void setUp() {
     @Test
     public void employeeService_sortByAge() {
-        Set<Employee> actualResult = employeeService.getEmployByOrder(employeesOfDifferentAges);
+        Set<Employee> actualResult = employeeService.getEmployeeByOrder(employeesOfDifferentAges);
```
This call invokes employeeService.getEmployeeByOrder(employeesOfDifferentAges) to verify ordering by age. This is correct and matches the task requirement to sort by age ascending. Consider adding an explicit test for `null` input (what should the service do when passed `null`?) if you want to specify the method contract more clearly.

mateacademy-ai-mentor commented on src/test/java/core/basesyntax/EmployeeServiceTest.java:
```diff
@@ -57,7 +57,7 @@ public void employeeService_sortByAge() {
     @Test
     public void employeeService_sortByName() {
-        Set<Employee> actualResult = employeeService.getEmployByOrder(employeesOfSameAges);
+        Set<Employee> actualResult = employeeService.getEmployeeByOrder(employeesOfSameAges);
```
This call verifies ordering by name when ages are equal (employeesOfSameAges). Good coverage for the secondary comparison requirement. You may also want to add a test where two employees share the same age and same name to document that duplicates are deduplicated by the comparator (SortedSet behavior).

mateacademy-ai-mentor commented on src/test/java/core/basesyntax/EmployeeServiceTest.java:
```diff
@@ -77,7 +77,7 @@ public void employeeService_sortByName() {
     @Test
     public void employeeService_sortByAgeAndName() {
-        Set<Employee> actualResult = employeeService.getEmployByOrder(employeesOfDifferentAgesAndNames);
+        Set<Employee> actualResult = employeeService.getEmployeeByOrder(employeesOfDifferentAgesAndNames);
```
This call tests combined ordering across a larger list (different ages and names). Good comprehensive check of comparator ordering across the dataset. Consider adding a case here (or a separate test) with an Employee whose `name` is `null` if you expect the comparator to be null-safe for names.

mateacademy-ai-mentor commented on src/test/java/core/basesyntax/EmployeeServiceTest.java:
```diff
@@ -117,7 +117,7 @@ public void employeeService_sortByAgeAndName() {
     @Test
     public void employeeService_emplyList() {
-        Set<Employee> actualResult = employeeService.getEmployByOrder(new ArrayList<>());
+        Set<Employee> actualResult = employeeService.getEmployeeByOrder(new ArrayList<>());
```
This call tests empty-list behavior (new ArrayList<>()). Good: it asserts size 0. As noted above, consider adding another test for `null` input to clarify expected behavior (throwing NPE vs. returning empty set).

mateacademy-ai-mentor reviewed (approved):
### Review Summary

**Decision: APPROVE**

Your implementation meets the core requirements: you created a custom comparator that orders by `age` ascending and then by `name` (using `Comparator.nullsFirst` for name), and you return a `TreeSet` constructed with that comparator from the exact required method `Set<Employee> getEmployeeByOrder(List<Employee> employees)`. This should pass the provided tests and reflects the intended behavior.

Minor improvements to consider (not blockers):
- Guard against `employees == null` before calling `addAll` to avoid a potential NPE. For example, return an empty `TreeSet` if the input is null.
- If the input list may contain `null` Employee elements, either filter them out before insertion or use `Comparator.nullsFirst(AGE_NAME_COMPARATOR)` on the `TreeSet`. Otherwise, document that null elements are not supported.
- Be aware that two Employees with the same age and name will be considered equal by the comparator and only one will remain in the `SortedSet` (deduplication). Document this if relevant.

Nice work using the fluent Comparator API and keeping the solution concise and readable.

***

✨ Please rate this review in the Platform's notification center[^1]. I will be grateful for your feedback! ✨

[^1]: [Rate AI review example](https://snipboard.io/gD6XKF.jpg)

koder95 committed:
Update src/main/java/core/basesyntax/EmployeeService.java
